### PR TITLE
feat(ui): Autocomplete for "IN" search queries

### PR DESCRIPTION
**Pull request by scttcper**
- after typing `key:[` it will automatically close the bracket so that its a valid query
- adds autocomplete for in queries using the AST

### CODE REVIEW

File: static/app/components/smartSearchBar/index.tsx [hunk 1]
```diff
@@ -565,6 +566,37 @@ class SmartSearchBar extends React.Component<Props, State> {
...
     }
+
+    const cursorToken = this.cursorToken;
+    if (
+      key === '[' &&
+      cursorToken?.type === Token.Filter &&
+      cursorToken.value.text.length === 0
+    ) {
+      const {query} = this.state;
+      if (isWithinToken(cursorToken.value, this.cursorPosition)) {
```
Comment (EvanPurkhiser) [inline static/app/components/smartSearchBar/index.tsx:577]: You can probably inline this `if` into the if above no?

File: static/app/components/smartSearchBar/index.tsx [hunk 2]
```diff
@@ -1221,12 +1274,33 @@ class SmartSearchBar extends React.Component<Props, State> {
...
-        // Include everything after the ':'
-        clauseStart = keyLocation.end.offset + 1;
-        clauseEnd = location.end.offset + 1;
-        replaceToken += ' ';
+        const valueToken = this.cursorValue ?? cursorToken.value;
+        const location = valueToken.location;
+
+        if (cursorToken.filter === FilterType.TextIn) {
+          // Current value can be null when adding a 2nd value
+          //             ▼ cursor
```
Comment (EvanPurkhiser) [inline static/app/components/smartSearchBar/index.tsx:1282]: not ↓ ? :trollface:

File: static/app/components/smartSearchBar/index.tsx [hunk 3]
```diff
@@ -565,6 +566,37 @@ class SmartSearchBar extends React.Component<Props, State> {
...
+        clauseEnd = location.end.offset + 1;
+        const beforeClause = query.substring(0, clauseStart);
+        let endClause = query.substring(clauseEnd);
+        // Add space before next clause if it exists
+        if (endClause) {
+          endClause = ` ${endClause}`;
+        }
+        const newQuery = `${beforeClause}${replaceToken}${endClause}`;
+        // Place cursor between inserted brackets
+        this.updateQuery(newQuery, beforeClause.length + replaceToken.length - 1);
```
Comment (EvanPurkhiser) [inline static/app/components/smartSearchBar/index.tsx:596]: We should maybe make a helper function for replacing text in a query. What do you think @davidenwang

File: static/app/components/smartSearchBar/index.tsx [hunk 4]
```diff
@@ -627,25 +659,16 @@ class SmartSearchBar extends React.Component<Props, State> {
...
-    const {parsedQuery} = this.state;
-
-    if (parsedQuery === null) {
-      return null;
-    }
-
-    const matchedTokens = [Token.Filter, Token.FreeText];
-    const cursor = this.cursorPosition;
+    const matchedTokens = [Token.Filter, Token.FreeText] as const;
+    return this.findTokensAtCursor(matchedTokens);
```
Comment (EvanPurkhiser) [inline static/app/components/smartSearchBar/index.tsx:663]: o yea i like dis

File: static/app/components/smartSearchBar/index.tsx [hunk 5]
```diff
@@ -1221,12 +1274,33 @@ class SmartSearchBar extends React.Component<Props, State> {
...
-        const location = cursorToken.value.location;
-        const keyLocation = cursorToken.key.location;
-        // Include everything after the ':'
-        clauseStart = keyLocation.end.offset + 1;
-        clauseEnd = location.end.offset + 1;
-        replaceToken += ' ';
+        const valueToken = this.cursorValue ?? cursorToken.value;
+        const location = valueToken.location;
+
+        if (cursorToken.filter === FilterType.TextIn) {
```
Comment (EvanPurkhiser) [inline static/app/components/smartSearchBar/index.tsx:1280]: What about `NumericIn` ?

**Review by evanpurkhiser — COMMENTED** [review]
Looking pretty good to me. Going to actually let @davidenwang take the final +1 on this one since it's his domain 😎

File: static/app/components/smartSearchBar/index.tsx (hunk 5 above)
Comment (scttcper) [reply static/app/components/smartSearchBar/index.tsx:1280]: we can already do `user.id:[123, 456]` with this change, maybe we can add it later

File: static/app/components/smartSearchBar/index.tsx (hunk 3 above)
Comment (davidenwang) [reply static/app/components/smartSearchBar/index.tsx]: yea that makes sense will remove the need to keep reconstructing the query

**Review by davidenwang — APPROVED** [review]
thanks for this scott!

**Pull request merged**